---
title: More SQL
duration: "1:25"
creator:
    name: Francesco Mosconi
    city: SF
---

# ![](https://ga-dash.s3.amazonaws.com/production/assets/logo-9f88ae6c9c3871690e33280fcf557f33.png) More SQL
Week 5 | Lesson 3.1


**Please do not push this notebook to your personal Github repo so we can keep database credentials in-house ;)**

### LEARNING OBJECTIVES
*After this lesson, you will be able to:*
- Sort results by a column using `ORDER BY`
- Simplify our syntax using aliases (`AS`)
- Match patterns using `LIKE`
- Select distinct items using `DISTINCT`
- Aggregate values using `GROUP BY`
- Filter on aggregations using `HAVING`


### STUDENT PRE-WORK
*Before this lesson, you should already be able to:*
- connect to a local or remote relational database
- add and remove data, create new tables, alter table schemas
- perform simple queries using SQL language

### INSTRUCTOR PREP
*Before this lesson, instructors will need to:*
- Read in / Review any dataset(s) & starter/solution code
- Generate a brief slide deck
- Prepare any specific materials
- Provide students with additional resources
- make sure the GA copy of the northwind database is available at:
        psql -h dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com -p 5432 -U dsi_student northwind
        password: gastudents


### LESSON GUIDE / REFERENCE
| TIMING  | TYPE  | TOPIC  |
|:-:|---|---|
| 5 mins | [Opening](#opening) | Opening |
| 10 mins | [Introduction](#introduction) | More SQL |
| 10 mins | [Guided-Reference](#guided-practice_1) | Guided Reference: ORDER BY |
| 10 mins | [Guided-Reference](#guided-practice_2) | Guided Reference: Aliases |
| 10 mins | [Guided-Reference](#guided-practice_3) | Guided Reference: SQL LIKE Operator |
| 10 mins | [Guided-Reference](#guided-practice_4) | Guided Reference: DISTINCT and LIMIT Operators |
| 10 mins | [Guided-Reference](#guided-practice_5) | Guided Reference: GROUP BY Operator |
| 10 mins | [Guided-Reference](#guided-practice_6) | Guided Reference: HAVING Operator |
| 10 mins | [Ind-practice](#ind-practice) | Independent Practice |
| 5 mins | [Conclusion](#conclusion) | Conclusion |

In [ ]:
%%bash

afplay /Users/davidyerrington/soundclips/love_boat.wav

# GROUP BY `table` Activity!

We are going to work together on the **Northwind** dataset to solidify our understanding of SQL basics.  We are going to split up into groups, and then we are going code for everyone like you wouldn't believe.

![](https://media.giphy.com/media/ZHlGzvZb130nm/giphy.gif)

Our activities will include:

- ** LIKE ** 
- ** GROUP** 
- ** DISTINCT and LIMIT**
- ** ORDER BY** 
- ** HAVING** (after we're comfortable with **`GROUP`** activities)

There will be a twist in this activity in which random person(s) will be selected from each group, each time, to present a solution on their given group topic, given a few minutes of deliberation and reference.  We will extend solutions based on actictivies frequenty so everyone has a chance to practice **and** discuss SQL implementation.

## First, Some Setup

In [1]:
# !pip install ipython-sql
%load_ext sql

/Users/sebozek/anaconda/lib/python2.7/site-packages/IPython/config.py:13: ShimWarning: The `IPython.config` package has been deprecated. You should import from traitlets.config instead.
  "You should import from traitlets.config instead.", ShimWarning)
/Users/sebozek/anaconda/lib/python2.7/site-packages/IPython/utils/traitlets.py:5: UserWarning: IPython.utils.traitlets has moved to a top-level traitlets package.
  warn("IPython.utils.traitlets has moved to a top-level traitlets package.")


### For Easy Reference

Here is a listing of tables in the **Northwind** database in our student Postgres database on **AWS**.

In [2]:
cfrom sqlalchemy import create_engine
import pandas as pd
engine = create_engine('postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com:5432/northwind')


pd.read_sql("SELECT * FROM pg_catalog.pg_tables WHERE schemaname='public'", con=engine)

schemaname             tablename tableowner tablespace hasindexes hasrules  \
0      public            categories        dsi       None       True    False   
1      public  customercustomerdemo        dsi       None       True    False   
2      public  customerdemographics        dsi       None       True    False   
3      public             customers        dsi       None       True    False   
4      public             employees        dsi       None       True    False   
5      public   employeeterritories        dsi       None       True    False   
6      public         order_details        dsi       None       True    False   
7      public                orders        dsi       None       True    False   
8      public              products        dsi       None       True    False   
9      public                region        dsi       None       True    False   
10     public              shippers        dsi       None       True    False   
11     public          shippers_tmp        dsi       None       True    False   
12     public             suppliers        dsi       None       True    False   
13     public           territories        dsi       None       True    False   
14     public              usstates        dsi       None      False    False   

   hastriggers  
0        False  
1        False  
2        False  
3        False  
4        False  
5        False  
6        False  
7        False  
8        False  
9        False  
10       False  
11       False  
12       False  
13       False  
14       False

In [3]:
# Special GROUP activity selector!
import random 
from subprocess import call

def select_group_activity(having=False, distinct=False, group=False):
    
    activities = ['LIKE', 'LIMIT', 'ORDER BY', 'ALIAS']
    
    if having:
        activities.append('HAVING')
    
    if group:
        activities.append('GROUP')
        
    if distinct:
        activities.append('DISTINCT')
    
    random_activity = random.choice(activities)
    call(["say", "Your selected activity is. %s!" % random_activity])
    return random_activity

In [4]:
# Because doesn't have `DESCRIBE [table]`
def describe_table(table="orders", connection=engine):
    
    sql = """
    SELECT
        column_name, data_type
    FROM
        information_schema.COLUMNS
    WHERE
        TABLE_NAME = '%s'
    """ % table

    return pd.read_sql(sql, con=connection)

In [5]:
## If you want to inspect the schema of a table, use this method

In [6]:
describe_table("orders", connection=engine)

column_name          data_type
0          OrderID           smallint
1       CustomerID          character
2       EmployeeID           smallint
3        OrderDate               date
4     RequiredDate               date
5      ShippedDate               date
6          ShipVia           smallint
7          Freight               real
8         ShipName  character varying
9      ShipAddress  character varying
10        ShipCity  character varying
11      ShipRegion  character varying
12  ShipPostalCode  character varying
13     ShipCountry  character varying

In [34]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com:5432/northwind

/* Select first ten items from orders */
/* SELECT * FROM orders LIMIT 10 */

/* Select three items after offeseting by five */
/* SELECT * FROM orders LIMIT 3 OFFSET 5 */

/* All ShipAddress with Commerce in it */
/* SELECT * FROM orders WHERE "ShipAddress" LIKE '%Commerce%' */

/* Select ShipAddress with Commerce in it and on the 28th */
/* SELECT * FROM orders WHERE "ShipAddress" LIKE '%Commerce%' AND CAST("ShippedDate" as varchar) LIKE '%28%' */


1 rows affected.


[(u'VICTE', u'Victuailles en stock')]

In [35]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com:5432/northwind

SELECT

"CustomerID" as Customer,
"ShipName" as Ship

FROM orders WHERE "ShipAddress" LIKE 
'%Commerce%' AND CAST("ShippedDate" as varchar) LIKE '%28%'
OFFSET 1

1 rows affected.


[(u'VICTE', u'Victuailles en stock')]

### Querying the database can be done 2 ways:

In [48]:
sql = """
SELECT * from employees WHERE "Notes" LIKE '%%Janet%%' 
"""

pd.read_sql(sql, con=engine)

EmployeeID   LastName FirstName                 Title TitleOfCourtesy  \
0           3  Leverling     Janet  Sales Representative             Ms.   

    BirthDate    HireDate             Address      City Region PostalCode  \
0  1963-08-30  1992-04-01  722 Moss Bay Blvd.  Kirkland     WA      98033   

  Country       HomePhone Extension Photo  \
0     USA  (206) 555-3412      3355    []   

                                               Notes  ReportsTo  \
0  Janet has a BS degree in chemistry from Boston...          2   

                                PhotoPath  
0  http://accweb/emmployees/leverling.bmp

In [59]:
sql = """
SELECT "RegionID" as "Region", COUNT(*) as "Region_Counts" FROM territories 
GROUP BY "RegionID"
"""

pd.read_sql(sql, con=engine)

Region  Region_Counts
0       1             19
1       2             15
2       3             11
3       4              8

In [62]:
sql = """
SELECT "FirstName", count(*) FROM employees 
GROUP BY "FirstName"
"""

pd.read_sql(sql, con=engine)

FirstName  count
0    Robert      1
1  Margaret      1
2     Nancy      1
3   Michael      1
4     Laura      1
5     Janet      1
6    Andrew      1
7    Steven      1
8      Anne      1

In [65]:
sql = """
SELECT "ShipCountry", "ShipCity", COUNT("ShipCity") FROM orders
GROUP BY "ShipCountry", "ShipCity"
ORDER BY count DESC
"""

pd.read_sql(sql, con=engine)

ShipCountry         ShipCity  count
0        Brazil   Rio de Janeiro     34
1            UK           London     33
2           USA            Boise     31
3        Brazil        Sao Paulo     31
4       Austria             Graz     30
5        Mexico      México D.F.     28
6       Germany        Cunewalde     28
7       Ireland             Cork     19
8        Sweden           Bräcke     19
9     Venezuela    San Cristóbal     18
10       Sweden            Luleå     18
11          USA      Albuquerque     18
12       France        Marseille     17
13    Argentina     Buenos Aires     16
14      Germany          München     15
15      Germany   Frankfurt a.M.     15
16      Finland             Oulu     15
17    Venezuela     Barquisimeto     14
18       Canada        Tsawassen     14
19          USA          Seattle     14
20      Germany      Brandenburg     14
21       France         Toulouse     14
22     Portugal           Lisboa     13
23       Canada         Montréal     13
24           UK       Colchester     13
25          USA         Portland     12
26        Italy    Reggio Emilia     12
27    Venezuela  I. de Margarita     12
28      Belgium        Charleroi     12
29       France       Strasbourg     11
..          ...              ...    ...
40      Germany        Stuttgart     10
41          USA           Lander      9
42       Brazil          Resende      9
43       Brazil         Campinas      9
44        Spain           Madrid      8
45  Switzerland             Bern      8
46      Belgium        Bruxelles      7
47      Finland         Helsinki      7
48       France           Nantes      7
49       Poland         Warszawa      7
50      Germany         Mannheim      7
51      Denmark        Kobenhavn      7
52       Norway          Stavern      6
53      Germany           Aachen      6
54      Germany           Berlin      6
55        Italy           Torino      6
56      Germany          Münster      6
57          USA            Elgin      5
58        Spain        Barcelona      5
59       France            Lille      5
60      Germany          Leipzig      5
61       France            Reims      5
62          USA    San Francisco      4
63       France            Paris      4
64       France       Versailles      4
65          USA         Kirkland      3
66       Canada        Vancouver      3
67          USA            Butte      3
68    Venezuela          Caracas      2
69          USA      Walla Walla      2

[70 rows x 3 columns]

In [72]:
sql = """
SELECT "ShipCountry", "ShipCity", COUNT("ShipCity") FROM orders
GROUP BY "ShipCountry", "ShipCity"
HAVING COUNT("ShipCity") > 20
ORDER BY count DESC
"""

pd.read_sql(sql, con=engine)

ShipCountry        ShipCity  count
0      Brazil  Rio de Janeiro     34
1          UK          London     33
2         USA           Boise     31
3      Brazil       Sao Paulo     31
4     Austria            Graz     30
5     Germany       Cunewalde     28
6      Mexico     México D.F.     28

In [74]:
sql = """
SELECT "ShipCountry", "ShipCity", COUNT("ShipCity") FROM orders
GROUP BY "ShipCountry", "ShipCity"
HAVING COUNT(*) BETWEEN 15 AND 35
"""

pd.read_sql(sql, con=engine)

ShipCountry        ShipCity  count
0      Finland            Oulu     15
1       Sweden          Bräcke     19
2          USA     Albuquerque     18
3       Brazil       Sao Paulo     31
4       Mexico     México D.F.     28
5       France       Marseille     17
6      Austria            Graz     30
7    Argentina    Buenos Aires     16
8      Ireland            Cork     19
9      Germany       Cunewalde     28
10     Germany         München     15
11          UK          London     33
12      Sweden           Luleå     18
13      Brazil  Rio de Janeiro     34
14   Venezuela   San Cristóbal     18
15     Germany  Frankfurt a.M.     15
16         USA           Boise     31

In [91]:
sql = """
SELECT "CustomerID", COUNT("CustomerID") as count FROM orders
WHERE "Freight" > 20
GROUP BY "CustomerID"
HAVING COUNT(*) = 29
"""

pd.read_sql(sql, con=engine)

CustomerID  count
0      SAVEA     29
1      ERNSH     29

In [93]:
sql = """
SELECT DISTINCT "RegionID" FROM territories """

pd.read_sql(sql, con=engine)

RegionID
0         1
1         2
2         3
3         4

In [95]:
sql = """
SELECT DISTINCT "ShipRegion" FROM orders"""

pd.read_sql(sql, con=engine)

ShipRegion
0            None
1           Essex
2        Co. Cork
3              MT
4         Táchira
5          Québec
6              NM
7   Nueva Esparta
8              AK
9              OR
10             BC
11             SP
12             CA
13             WA
14             RJ
15             DF
16             WY
17           Lara
18  Isle of Wight
19             ID

In [97]:
sql = """
SELECT o.* FROM orders o"""

pd.read_sql(sql, con=engine)

OrderID CustomerID  EmployeeID   OrderDate RequiredDate ShippedDate  \
0      10248      VINET           5  1996-07-04   1996-08-01  1996-07-16   
1      10249      TOMSP           6  1996-07-05   1996-08-16  1996-07-10   
2      10250      HANAR           4  1996-07-08   1996-08-05  1996-07-12   
3      10251      VICTE           3  1996-07-08   1996-08-05  1996-07-15   
4      10252      SUPRD           4  1996-07-09   1996-08-06  1996-07-11   
5      10253      HANAR           3  1996-07-10   1996-07-24  1996-07-16   
6      10254      CHOPS           5  1996-07-11   1996-08-08  1996-07-23   
7      10255      RICSU           9  1996-07-12   1996-08-09  1996-07-15   
8      10256      WELLI           3  1996-07-15   1996-08-12  1996-07-17   
9      10257      HILAA           4  1996-07-16   1996-08-13  1996-07-22   
10     10258      ERNSH           1  1996-07-17   1996-08-14  1996-07-23   
11     10259      CENTC           4  1996-07-18   1996-08-15  1996-07-25   
12     10260      OTTIK           4  1996-07-19   1996-08-16  1996-07-29   
13     10261      QUEDE           4  1996-07-19   1996-08-16  1996-07-30   
14     10262      RATTC           8  1996-07-22   1996-08-19  1996-07-25   
15     10263      ERNSH           9  1996-07-23   1996-08-20  1996-07-31   
16     10264      FOLKO           6  1996-07-24   1996-08-21  1996-08-23   
17     10265      BLONP           2  1996-07-25   1996-08-22  1996-08-12   
18     10266      WARTH           3  1996-07-26   1996-09-06  1996-07-31   
19     10267      FRANK           4  1996-07-29   1996-08-26  1996-08-06   
20     10268      GROSR           8  1996-07-30   1996-08-27  1996-08-02   
21     10269      WHITC           5  1996-07-31   1996-08-14  1996-08-09   
22     10270      WARTH           1  1996-08-01   1996-08-29  1996-08-02   
23     10271      SPLIR           6  1996-08-01   1996-08-29  1996-08-30   
24     10272      RATTC           6  1996-08-02   1996-08-30  1996-08-06   
25     10273      QUICK           3  1996-08-05   1996-09-02  1996-08-12   
26     10274      VINET           6  1996-08-06   1996-09-03  1996-08-16   
27     10275      MAGAA           1  1996-08-07   1996-09-04  1996-08-09   
28     10276      TORTU           8  1996-08-08   1996-08-22  1996-08-14   
29     10277      MORGK           2  1996-08-09   1996-09-06  1996-08-13   
..       ...        ...         ...         ...          ...         ...   
800    11048      BOTTM           7  1998-04-24   1998-05-22  1998-04-30   
801    11049      GOURL           3  1998-04-24   1998-05-22  1998-05-04   
802    11050      FOLKO           8  1998-04-27   1998-05-25  1998-05-05   
803    11051      LAMAI           7  1998-04-27   1998-05-25        None   
804    11052      HANAR           3  1998-04-27   1998-05-25  1998-05-01   
805    11053      PICCO           2  1998-04-27   1998-05-25  1998-04-29   
806    11054      CACTU           8  1998-04-28   1998-05-26        None   
807    11055      HILAA           7  1998-04-28   1998-05-26  1998-05-05   
808    11056      EASTC           8  1998-04-28   1998-05-12  1998-05-01   
809    11057      NORTS           3  1998-04-29   1998-05-27  1998-05-01   
810    11058      BLAUS           9  1998-04-29   1998-05-27        None   
811    11059      RICAR           2  1998-04-29   1998-06-10        None   
812    11060      FRANS           2  1998-04-30   1998-05-28  1998-05-04   
813    11061      GREAL           4  1998-04-30   1998-06-11        None   
814    11062      REGGC           4  1998-04-30   1998-05-28        None   
815    11063      HUNGO           3  1998-04-30   1998-05-28  1998-05-06   
816    11064      SAVEA           1  1998-05-01   1998-05-29  1998-05-04   
817    11065      LILAS           8  1998-05-01   1998-05-29        None   
818    11066      WHITC           7  1998-05-01   1998-05-29  1998-05-04   
819    11067      DRACD           1  1998-05-04   1998-05-18  1998-05-06   
820    11068      QUEEN           8  1998-05-04   199

## Inline with the sql notebook extension

In [ ]:
%%sql postgresql://dsi_student:gastudents@dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com/northwind
        
SELECT * FROM orders WHERE "RequiredDate" > '1996-07-01' LIMIT 10

<a name="opening"></a>
## Opening (5 mins)
We have seen how to connect to a local sqlite database and to a remote postgresql database.

**Check:** What SQL commands have we learned so far?

> Answer:
- CREATE
- INSERT
- DELETE
- UPDATE
- SELECT

**Check:** What different commands have we learned for SQLite and PostgreSQL?

> Answer:
- how to list schema and tables


<a name="introduction"></a>
## More SQL (10 mins)

In this lecture we'll learn a few more useful `SQL` commands that will give us the ability to perform more complex queries.

In particular we will learn to:
- Sort results by a column using `ORDER BY`
- Simplify our syntax using aliases
- Match patterns using `LIKE`
- Select distinct items using `DISTINCT`
- Aggregate values using `GROUP BY`
- Filter on aggregations using `HAVING`

For all the following example we will use the well-known [Northwind sample database](https://northwinddatabase.codeplex.com/).

We have imported this database into our PostgreSQL instance that you can find connecting here:

    psql -h dsi.c20gkj5cvu3l.us-east-1.rds.amazonaws.com -p 5432 -U dsi_student northwind
    password: gastudents

We will use a few tables from this database, here is what they look like:

`customers`:

|CustomerID |CompanyName |ContactName | ContactTitle |Address|City | Region | PostalCode | Country |Phone | Fax|
|---|
|ALFKI| Alfreds Futterkiste| Maria Anders | Sales Representative | Obere Str. 57 | Berlin|| 12209| Germany | 030-0074321| 030-0076545|
|ANATR| Ana Trujillo Emparedados y helados | Ana Trujillo | Owner| Avda. de la Constitución 2222 | México D.F. || 05021| Mexico| (5) 555-4729 | (5) 555-3745|
|ANTON| Antonio Moreno Taquería| Antonio Moreno | Owner| Mataderos2312 | México D.F. || 05023| Mexico| (5) 555-3932 |
|...|...|...|...|...|...|...|...|...|...|...|

`orders`:

|OrderID | CustomerID | EmployeeID | OrderDate| RequiredDate | ShippedDate | ShipVia | Freight | ShipName|ShipAddress |ShipCity| ShipRegion | ShipPostalCode | ShipCountry |
|----|
|10248 | VINET|5 | 1996-07-04 | 1996-08-01 | 1996-07-16| 3 | 32.38 | Vins et alcools Chevalier | 59 rue de l'Abbaye | Reims|| 51100| France|
|10249 | TOMSP|6 | 1996-07-05 | 1996-08-16 | 1996-07-10| 1 | 11.61 | Toms Spezialitäten| Luisenstr. 48| Münster|| 44087| Germany|
|10250 | HANAR|4 | 1996-07-08 | 1996-08-05 | 1996-07-12| 2 | 65.83 | Hanari Carnes | Rua do Paço, 67| Rio de Janeiro | RJ | 05454-876| Brazil|
|...|...|...|...|...|...|...|...|...|...|...|

`order_details`:

| OrderID |  ProductID |  UnitPrice | Quantity | Discount |
| ----- |
|10248|11|14|12|0|
|10248|42|9.8|10|0|
|10248|72|34.8|5|0|
|10249|14|18.6|9|0|
|10249|51|42.4|40|0|
|10250|41|7.7|10|0|
|...|...|...|...|...|


**Check** Let's come up with a few queries you'd be curious to run on these tables.


<a name="guided-practice_1"></a>
## Guided Practice: ORDER BY (10 mins)

The `ORDER BY` keyword is used to sort the result-set by one or more columns. It sorts the records in ascending order by default. To sort the records in a descending order, you can use the `DESC` keyword.

### SQL ORDER BY Syntax

```sql
SELECT _column_name_,_ column_name_  
FROM _table_name_  
ORDER BY _column_name _ASC|DESC,_ column_name_ ASC|DESC;
```

#### Example

The following SQL statement selects all customers from the "Customers" table,  sorted by the "Country" column:

```sql
SELECT "CustomerID", "CompanyName", "Country" FROM customers
ORDER BY "Country";
```

|CustomerID|CompanyName|Country|
|----|
|OCEAN|Océano Atlántico Ltda.|Argentina|
|CACTU|Cactus Comidas para llevar|Argentina|
|RANCH|Rancho grande |Argentina|
|...|...|...|


#### ORDER BY DESC Example

The following SQL statement selects all customers from the "Customers" table,  sorted DESCENDING by the "Country" column:


```sql
SELECT "CustomerID", "CompanyName", "Country" FROM customers
ORDER BY "Country" DESC;
```

|CustomerID|CompanyName|Country|
|----|
|GROSR|GROSELLA-Restaurante |Venezuela|
|LILAS|LILA-Supermercado|Venezuela|
|HILAA|HILARION-Abastos |Venezuela|
|...|...|...|



#### ORDER BY Several Columns Example

The following SQL statement selects all customers from the `customers` table,  sorted by the "Country" and the "CustomerName" column:

```sql
SELECT "CustomerID", "CompanyName", "Country" FROM customers
ORDER BY "Country", "CompanyName";
```

|CustomerID|CompanyName|Country|
|----|
|CACTU|Cactus Comidas para llevar|Argentina|
|OCEAN|Océano Atlántico Ltda.|Argentina|
|RANCH|Rancho grande |Argentina|
|...|...|...|

**Check** run a few queries on the provided database and try using the `ORDER BY` command.

<a name="guided-practice_2"></a>
## Guided Practice: Aliases (10 mins)

SQL aliases are used to give a database table, or a column in a table, a temporary name. Basically aliases are created to make column names more readable.

### SQL Alias Syntax for Columns

```sql
SELECT _column_name_ AS _alias_name_  
FROM _table_name;_
```
### SQL Alias Syntax for Tables
```sql
SELECT _column_name(s)_  
FROM _table_name _AS _alias_name;_
```

#### Alias Examples for Table Columns

The following SQL statement specifies two aliases, one for the `CompanyName`  column and one for the `ContactName` column:

```sql
SELECT "CompanyName" AS "Customer", "ContactName" AS "[Contact Person]"
FROM customers;
```

|Customer|[Contact Person]|
|----|
|Alfreds Futterkiste| Maria Anders|
|Ana Trujillo Emparedados y helados| Ana Trujillo|
|Antonio Moreno Taquería| Antonio Moreno|
|...|...|


In the following SQL statement we combine four columns (Address, City, PostalCode and Country) and create an alias named `Address`:

```sql
SELECT "CompanyName" AS "Customer",
    CONCAT("Address",', ',"City",', ', "PostalCode",', ',"Country") AS "Address"
FROM customers;
```

|Customer|Address|
|----|
|Alfreds Futterkiste| Obere Str. 57, Berlin, 12209, Germany|
|Ana Trujillo Emparedados y helados| Avda. de la Constitución 2222, México D.F., 05021, Mexico|
|Antonio Moreno Taquería Mataderos| 2312, México D.F., 05023, Mexico|
|...|...|


#### Alias Example for Tables

The following SQL statement selects all the orders from the customer with  CustomerID=4 (Around the Horn). We use the `customers` and `orders` tables, and  give them the table aliases of `c` and `o` respectively (Here we have used  aliases to make the SQL shorter):


```sql
SELECT o."OrderID", o."OrderDate", c."CompanyName"
FROM customers AS c, orders  AS o  
WHERE c."CompanyName" = 'Around the Horn' AND c."CustomerID"=o."CustomerID";
```

|CustomerID|OrderDate|CompanyName|
|----|
|10355|1996-11-15|Around the Horn|
|10383|1996-12-16|Around the Horn|
|10453|1997-02-21|Around the Horn|
|10558|1997-06-04|Around the Horn|
|...|...|...|

**Check** what would be the same SQL statement without aliases?

Aliases can be useful when:

- There are more than one table involved in a query
- Functions are used in the query
- Column names are big or not very readable
- Two or more columns are combined together

**Check** try running the same queries as above with aliases. Are they more readable?

<a name="guided-practice_3"></a>
## Guided Practice: SQL LIKE Operator (10 mins)

The LIKE operator is used in a WHERE clause to search for a specified pattern in a column.

### SQL LIKE Syntax
```sql
SELECT _column_name(s)_  
FROM _table_name_  
WHERE _column_name_ LIKE _pattern_;
```

#### SQL LIKE Operator Examples

The following SQL statement selects all customers with a City starting with  the letter "S":

```sql
SELECT * FROM customers
WHERE "City" LIKE 'S%';
```

|CustomerID |CompanyName |ContactName | ContactTitle |Address|City | Region | PostalCode | Country |Phone | Fax|
|---|
|BLONP| Blondesddsl père et fils|Frédérique Citeaux|Marketing Manager| 24, place Kléber|Strasbourg||67000| France|88.60.15.31|88.60.15.32|
|COMMI| Comércio Mineiro|Pedro Afonso|Sales Associate|Av. dos Lusíadas, 23|Sao Paulo| SP|05432-043| Brazil|(11) 555-7647| 
|FAMIA| Familia Arquibaldo|Aria Cruz| Marketing Assistant | Rua Orós, 92|Sao Paulo| SP|05442-030| Brazil|(11) 555-9857| 
|...|...|...|...|...|...|...|...|...|...|...|

**Tip:** The "%" sign is used to define wildcards (missing  letters) both before and after the pattern. Also notice that `PostgreSQL` is case sensitive.

The following SQL statement selects all customers with a City ending with the  letter "s":

```sql
SELECT * FROM customers
WHERE "City" LIKE '%s';
```

|CustomerID |CompanyName |ContactName | ContactTitle |Address|City | Region | PostalCode | Country |Phone | Fax|
|---|
|CACTU| Cactus Comidas para llevar|Patricio Simpson|Sales Agent|Cerrito 333 |Buenos Aires||1010|Argentina| (1) 135-5555|(1) 135-4892|
|DUMON| Du monde entier|Janine Labrune|Owner| 67, rue des Cinquante Otages|Nantes||44000| France|40.67.88.88| 40.67.89.89|
|FRANR| France restauration|Carine Schmitt|Marketing Manager| 54, rue Royale|Nantes||44000| France|40.32.21.21| 40.32.21.20|
|...|...|...|...|...|...|...|...|...|...|...|

Using the NOT keyword allows you to select records that do NOT match the pattern. The following SQL statement selects all customers with Country NOT  containing the pattern "land":

```sql
SELECT * FROM customers
WHERE "Country" NOT LIKE '%land%';
```
|CustomerID |CompanyName |ContactName | ContactTitle |Address|City | Region | PostalCode | Country |Phone | Fax|
|---|
|ALFKI| Alfreds Futterkiste|Maria Anders|Sales Representative|Obere Str. 57| Berlin||12209| Germany|030-0074321| 030-0076545|
|ANATR| Ana Trujillo Emparedados y helados|Ana Trujillo|Owner| Avda. de la Constitución 2222| México D.F.|| 05021| Mexico|(5) 555-4729|(5) 555-3745|
|ANTON| Antonio Moreno Taquería|Antonio Moreno|Owner| Mataderos|2312 México D.F.|| 05023| Mexico|(5) 555-3932|
|...|...|...|...|...|...|...|...|...|...|...|

**Check** run a few queries on the provided database and try using the `LIKE` command.

<a name="guided-practice_4"></a>
## Guided Practice: DISTINCT and LIMIT Operators (10 mins)

The SELECT DISTINCT statement is used to return only distinct (different) values. In a table, a column may contain many duplicate values; and sometimes you only want to list the different (distinct) values.

### SQL SELECT DISTINCT Syntax
```sql
SELECT DISTINCT _column_name_,_column_name_  
FROM _table_name_;
```

#### SELECT DISTINCT Example

The following SQL statement selects only the distinct values from the "City"  columns from the "Customers" table:

```sql
SELECT DISTINCT "City" FROM customers;
```

| City|
|---|
|Leipzig|
|London|
|Nantes|
|...|

The `DISTINCT` operator is equivalent to the `numpy.unique` command.

### SQL SELECT LIMIT Syntax
Sometimes we may want to only retrieve a fixed number of records from the database. This is where the `LIMIT` operator comes in.

```sql
SELECT _column_name_,_column_name_  
FROM _table_name_
LIMIT _number_of_records;
```

#### SELECT LIMIT Example

The following SQL statement selects only the first 3 values from the "City"  columns from the "Customers" table:

```sql
SELECT DISTINCT "City"
FROM customers
LIMIT 3;
```

| City|
|---|
|Leipzig|
|London|
|Nantes|

The `LIMIT` operator is equivalent to the `pandas.Dataframe.head` command.

**Check** try running the same queries using `DISTINCT` and `LIMIT`

<a name="guided-practice_5"></a>
## Guided Practice: GROUP BY Operator (10 mins)

A table may contain several records that have a common key. Consider for example the following `order_details` table:


| OrderID |  ProductID |  UnitPrice | Quantity | Discount |
| ----- |
|10248|11|14|12|0|
|10248|42|9.8|10|0|
|10248|72|34.8|5|0|
|10249|14|18.6|9|0|
|10249|51|42.4|40|0|
|10250|41|7.7|10|0|
|...|...|...|...|...|


The _GROUP BY_ statement is used in conjunction with the aggregate functions to group the result-set by one or more columns. For example we may want to know the total number of items purchased in each order.

#### SQL GROUP BY Syntax

The syntax to perform a _GROUP BY_ operation is the following:

```sql
SELECT column_name, aggregate_function(column_name)  
FROM table_name  
WHERE column_name operator value  
GROUP BY column_name;
```
We select the column we want to aggregate on and the function.


#### Example

In order to calculate the total amount of items in a specific order we can use the following query:

```sql
SELECT "OrderID", SUM("Quantity")
FROM order_details
GROUP BY "OrderID"
```

Note that in `PostgreSQL` unquoted names are case-insensitive. Thus `SELECT * FROM hello` and `SELECT * FROM HELLO` are equivalent.

However, quoted names are case-sensivite. `SELECT * FROM "hello"` is not equivalent to `SELECT * FROM "HELLO"`.

To make a "bridge" between quoted names and unquoted names, unquoted names are implicitly lowercased, thus hello, HELLO and HeLLo are equivalent to "hello", but not to "HELLO" or "HeLLo".

Thus, when creating entities (tables, views, procedures, etc) in PostgreSQL, you should specify them either unquoted, or quoted-but-lowercased.

<a name="guided-practice_6"></a>
## Guided Practice: HAVING Operator (10 mins)

The HAVING clause was added to SQL because the WHERE keyword could not be used with aggregate functions. Let's consider the previous example where we aggregated the `Quantity` of an order. What if we would like to filter the results for orders where the Maximum discout applied on any item is lower than 10%?

#### SQL HAVING Syntax

```sql
SELECT column_name, aggregate_function(column_name)  
FROM table_name  
WHERE column_name operator value  
GROUP BY column_name  
HAVING aggregate_function(column_name) operator value;
```

##### Example

```sql
SELECT "OrderID", SUM("Quantity"), MAX("Discount")
FROM order_details
GROUP BY "OrderID"
HAVING MAX("Discount") <= 0.1;
```

|OrderID|sum|max|
|---|
|10501|20|0|
|10295|4|0|
|10827|36|0|

<a name="ind-practice"></a>
## Independent Practice (10 minutes)

Now that you've learned a lot more commands, try combining them:

- retrieve the top 3 orders by number of items ordered
- retrieve the oldest 5 orders
- retrieve the 5 most recent orders. Nest that query into another one that ranks these 5 orders by shipping country (alphabetical order)
- try some queries on your own

> Answers:
```sql
SELECT "OrderID", SUM("Quantity") AS "Sum"
FROM order_details
GROUP BY "OrderID"
ORDER BY "Sum" DESC
LIMIT 3;
```
```sql
SELECT "OrderID", "OrderDate"
FROM orders
ORDER BY "OrderDate"
LIMIT 5;
```
```sql
SELECT * FROM (
	SELECT "OrderID", "OrderDate", "ShipCountry"
	FROM orders
	ORDER BY "OrderDate" DESC
	LIMIT 5
	) AS fivemostrecent
ORDER BY "ShipCountry";
```

<a name="conclusion"></a>
## Conclusion (5 mins)
In this lesson we have learned many more new commands to make our SQL queries more powerful.


In particular we have learned how to:
- Sort results by a column using `ORDER BY`
- Simplify our syntax using aliases
- Match patterns using `LIKE`
- Select distinct items using `DISTINCT`
- Aggregate values using `GROUP BY`
- Filter on aggregations using `HAVING`

**Check** can you think of a few more business cases where these are useful?

***

### ADDITIONAL RESOURCES

- [PostgreSQL Documentation](http://www.postgresql.org/docs/)
- [Sqlite Documentation](https://www.sqlite.org/docs.html)